### PR TITLE
[autograd.Function] Add docs on the functorch interaction

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91452

This PR:
- Updates autograd.Function.forward docs to reflect how you either
  define a forward with ctx or a separate forward and setup_context
- Updates the "Extending Autograd" docs to suggest the usage of
  autograd.Function with separate forward and setup_context. This should
  be the default because there is a low barrier to go from this to
  an autograd.Function that is fully supported by functorch transforms.
- Adds a new "Extending torch.func with autograd.Function" doc that
  explains how to use autograd.Function with torch.func. It also
  explains how to use generate_vmap_rule and how to manually write a
  vmap staticmethod.

While writing this, I noticed that the implementation of
setup_context staticmethod/generate_vmap_rule/vmap staticmethod are a
bit inconsistent with the other method/attributes on autograd.Function:
- https://github.com/pytorch/pytorch/issues/91451
- I'm happy to fix those if we think it is a problem, either in this PR
  or a followup (this PR is getting long, I want some initial docs
  out that I can point early adopters at, and fixing the problems in the
  future isn't really BC-breaking).

Test Plan:
- view docs preview

### CODE REVIEW

File: torch/autograd/function.py [hunk 1]
```diff
@@ -296,12 +296,36 @@ def __init__(cls, name, bases, attrs):
...
+
+            @staticmethod
+            def setup_context(ctx: Any, inputs: Tuple[Any, ...], output: Any) -> None:
+                pass
 
-        This function is to be overridden by all subclasses.
 
-        It must accept a context ctx as the first argument, followed by any
-        number of arguments (tensors or other types).
+        - You must also define a setup_context staticmethod to handle setting up the
```
Comment (soulitzer) [inline torch/autograd/function.py:324]: nit: maybe add a point about: "forward no longer takes context ctx as the first argument."

File: docs/source/notes/extending.rst [hunk 2]
```diff
@@ -72,6 +73,12 @@ Take the following steps:
   tensors if there are multiple outputs. Also, please refer to the
   docs of :class:`Function` to find descriptions of useful methods that can be
   called only from :meth:`~Function.forward`.
+- ``setup_context`` (optional). One can either write a "combined" :meth:`~Function.forward` that
+  accepts a ``ctx`` object or (as of PyTorch 2.0) a separate :meth:`~Function.forward` that does
+  not accept ``ctx`` and a ``setup_context`` method where the ``ctx`` modification happens.
+  In general the separate :meth:`~Function.forward` and ``setup_context`` is closer to how
+  PyTorch native operations work and therefore more composable with various PyTorch subsystems.
+  See :ref:`combining-forward-context` for more details
```
Comment (soulitzer) [inline docs/source/notes/extending.rst:81]: We mention that we shouldn't do compute inside setup_context in an example somewhere above, but maybe we should mention that here too.

File: docs/source/notes/extending.func.rst [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+So you'd like to use :class:`torch.autograd.Function` with the :mod:`torch.func`
+transforms like :func:`torch.vmap`, :func:`torch.func.grad`, etc.
+
+There are two main use cases:
+
+- you wish to call code that does not contain PyTorch operations and
+  have it work with function transforms. That is, the :class:`torch.autograd.Function`'s
+  forward/backward/etc calls into functions from other systems like C++, CUDA, numpy.
+- you wish to specify custom gradient rules, like
+  JAX's `custom_vjp/custom_jvp <https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html>`_
```
Comment (soulitzer) [inline docs/source/notes/extending.func.rst:17]: Do we want to mention that there might be unexpected behavior if the user tries to mix the two use cases. For example if user specifies a custom vmap rule AND wants to do custom gradients, the former can override the latter.

File: docs/source/notes/extending.rst [hunk 4]
```diff
@@ -233,6 +289,54 @@ See :ref:`grad-check` for more details on finite-difference gradient comparisons
...
+
+There are two main ways to define :class:`~Function`. Either:
+
+- define a :meth:`~Function.forward` that combines the forward compute logic with ``setup_context``
+- (as of PyTorch 2.0) define a separate :meth:`~Function.forward` and ``setup_context``.
+
+We recommend the second option (separate :meth:`~Function.forward` and ``setup_context``)
+because that is closer to how PyTorch native operations are implemented and it composes
+with :mod:`torch.func` transforms. However, we plan to support both approaches going forward;
+combining :meth:`~Function.forward` with ``setup_context`` leads to more flexibility.
```
Comment (soulitzer) [inline docs/source/notes/extending.rst:305]: nit: we aren't super specific about why combining is more flexible here (i.e. allow for saving intermediates?)

File: docs/source/notes/extending.func.rst [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+should behave under :func:`torch.vmap`. That is, it defines how to transform
+the :meth:`~Function.forward` to run over inputs with an additional dimension (the dimension
+being vmapped over). This is similar to how :func:`torch.vmap` is implemented over
+PyTorch operations: for each operation, we define a vmap rule (sometimes also
+referred to as a "batching rule").
+
+Here's how to define the ``vmap`` staticmethod:
+
+- the signature is ``vmap(info, in_dims: Tuple[Optional[int]], *args)``, where
+  ``*args`` is the same as the args to :meth:`~Function.forward`.
```
Comment (soulitzer) [inline docs/source/notes/extending.func.rst:337]: should vmap take `**kwargs` for consistency since forward does (though they have to be passed positionally because apply doesn't take true kwargs)

File: docs/source/notes/extending.func.rst [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+- The vmap staticmethod is responsible for defining how the :meth:`~Function.forward` should behave
+  under :func:`torch.vmap`. That is, given inputs with an additional dimension
+  (specified by ``in_dims``), how do we compute the batched version of :meth:`~Function.forward`?
+- For each arg in ``args``, ``in_dims`` has a corresponding ``Optional[int]``.
+  It is ``None`` if the arg is not a Tensor or if the arg is not being vmapped over,
+  otherwise, it is an integer specifying what dimension of the Tensor is being vmapped
+  over.
+- ``info`` is a collection of additional metadata that may be helpful. For example,
```
Comment (soulitzer) [inline docs/source/notes/extending.func.rst:345]: Do we want to mention all the metadata here? (or link to somewhere where one could find that information)

**Review by soulitzer — APPROVED** [review]
Nice!

**Comment by soulitzer** [comment]
Will take another look once the docs render

**Comment by zou3519** [comment]
> Will take another look once the docs render

Thanks!

File: docs/source/notes/extending.func.rst (hunk 6 above)
Comment (zou3519) [reply docs/source/notes/extending.func.rst:345]: Sure

File: docs/source/notes/extending.rst (hunk 4 above)
Comment (zou3519) [reply docs/source/notes/extending.rst:305]: yeah, saving intermediates without promoting them to be outputs

File: docs/source/notes/extending.func.rst (hunk 3 above)
Comment (zou3519) [reply docs/source/notes/extending.func.rst:17]: Is that unexpected behavior though?

If I'm a user, and I write a forward() + backward() using only PyTorch operations, and I want custom vmap behavior, then I would write a custom vmap rule rather than use generate_vmap_rule.

Though I agree we should mention that this case is possible. I'll put it below in the "torch.vmap Support" section.

File: docs/source/notes/extending.rst [hunk 7]
```diff
@@ -205,6 +226,41 @@ And here, we optimize the above example by calling set_materialize_grads(False):
             # Gradients of non-Tensor arguments to forward must be None.
             return grad_output * ctx.constant, None
 
+If you need any "intermediate" Tensors computed in :meth:`~Function.forward` to be saved,
```
[the remark this answers is not in the document]
Comment (zou3519) [reply docs/source/notes/extending.rst:229]: Sgtm

File: docs/source/notes/extending.func.rst (hunk 5 above)
Comment (zou3519) [reply docs/source/notes/extending.func.rst:337]: I might take this as a follow-up item. I am not sure:
- `setup_context(ctx, inputs, output)` takes in `inputs` that were directly passed to forward
- If there were ANY unspecified values that are default kwargs, then we have a problem -- they're not passed to setup_context, so setup_context doesn't get their default value (if it needs to save them for backward)
- So the conclusion is that autograd.Function with setup_context has some really awkward handling for default kwargs.

I'm inclined to say in the docs that autograd.Function with setup_context doesn't support `**kwargs`

File: docs/source/notes/extending.func.rst (hunk 3 above)
Comment (soulitzer) [reply docs/source/notes/extending.func.rst:17]: yeah nah, not unexpected behavior, but probably tricky enough to deserve a mention

Comment (zou3519) [reply docs/source/notes/extending.func.rst:17]: Added a mention, as well as a small note about the grad(vmap(Func)) semantics

File: docs/source/notes/extending.func.rst (hunk 5 above)
Comment (zou3519) [reply docs/source/notes/extending.func.rst:337]: I end up changing the documentation to suggest that default args and keywords args be passed via wrapper so it's clearer how to handle them, but not changing the suggested signature of the vmap staticmethod

**Comment by zou3519** [comment]
Docs preview is up if folks wanted to take another pass: https://docs-preview.pytorch.org/91452/notes/extending.func.html
